code/json.sls

```diff
@@ -30,172 +30,185 @@
 
 (library (json)
   (export json-write
-	  json-read)
+          json-read)
   
   (import
-    (except (rnrs) define-record-type error)
-    (rnrs r5rs)
-    (rnrs mutable-pairs)
-    (only (srfi :1 lists) lset-union append-map fold)
-    (srfi :6 basic-string-ports)
-    (srfi :9 records)
-    (srfi :23 error)
-    (only (chezscheme) include)
-    (packrat))
+   (except (rnrs) define-record-type error)
+   (rnrs r5rs)
+   (rnrs mutable-pairs)
+   (only (srfi :1 lists) lset-union append-map fold iota)
+   (srfi :6 basic-string-ports)
+   (srfi :9 records)
+   (srfi :23 error)
+   (only (chezscheme) include)
+   (packrat))
 
   (define json-write
     (let ()
+      (define indent-level 0)
+      (define (display-level)
+        (display
+         (apply string-append (map (lambda (_) "  ") (iota indent-level)))))
       (define (write-ht vec p)
-	(display "{" p)
-	(do ((need-comma #f #t)
-	     (vec vec (cdr vec)))
-	    ((null? vec))
-	  (if need-comma
-	      (display ", " p)
-	      (set! need-comma #t))
-	  (let* ((entry (car vec))
-		 (k (car entry))
-		 (v (cdr entry)))
-	    (cond
-	      ((symbol? k) (write (symbol->string k) p))
-	      ((string? k) (write k p)) ;; for convenience
-	      (else (error "Invalid JSON table key in json-write" k)))
-	    (display ": " p)
-	    (write-any v p)))
-	(display "}" p))
+        (display "{\n" p)
+        (set! indent-level (+ indent-level 1))
+        (do ((need-comma #f #t)
```

Comment:
Configlet outputs indents with tabstop 2. Probably makes sense to increment/decrement indent-level by 2

Reply:
I was isolating the indent size  to its own variable, so that indent level more acurately freflects the actual level of indentation (not the number of spaces).

Does that logic sound good to you?

Reply:
So, I made the `tabstop-size` variable.  I think that is easier.

Although I dislike that it cannot be changed on the fly by a user of the library.

Part of what I would like to do with this library is pull the functions out of `json-write`, and give them a feature to print either pretty or not.  That way we could eliminate the excessive repetition the library has now.

When I do that I will likely try using an optional tabstop-size argument.

Reply:
You could, though I don't think the repetition here is overly harmful. It's probably simpler to just have two json printing functions. To get optional `tabstop` argument you can define with `case-lambda`

Reply:
[https://www.scheme.com/tspl4/binding.html#./binding:h3](https://www.scheme.com/tspl4/binding.html#./binding:h3)